android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt

```diff
@@ -0,0 +1,162 @@
+import org.mockito.Mockito.`when`
+import org.mockito.Mockito.spy
+import org.mockito.MockitoAnnotations
+import org.mockito.kotlin.anyOrNull
+import org.mockito.kotlin.argumentCaptor
+import org.mockito.kotlin.doAnswer
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+import org.robolectric.Robolectric
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+@ObsoleteCoroutinesApi
+@ExperimentalCoroutinesApi
+class BrowserSignInFlowTest {
+    private val mainThreadSurrogate = newSingleThreadContext("UI thread")
+
+    @Mock
+    private lateinit var logtoConfig: LogtoConfig
+
+    @Mock
+    private lateinit var logtoApiClient: LogtoApiClient
+
+    @Mock
+    private lateinit var onComplete: (exception: LogtoException?, tokenSet: TokenSet?) -> Unit
+
+    private val logtoExceptionCaptor = argumentCaptor<LogtoException>()
+    private val tokenSetCaptor = argumentCaptor<TokenSet>()
+
+    @Before
+    fun setUp() {
+        Dispatchers.setMain(mainThreadSurrogate)
+        MockitoAnnotations.openMocks(this)
+        `when`(logtoConfig.redirectUri).thenReturn("redirectUri")
+    }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
+        mainThreadSurrogate.close()
+    }
+
+    @Test
+    @Suppress("UNCHECKED_CAST")
+    fun shouldStartSignIn() {
+        val oidcConfiguration: OidcConfiguration = mock()
+        `when`(oidcConfiguration.authorizationEndpoint).thenReturn("authorizationEndpoint")
+
+        doAnswer {
+            val block = it.arguments[0] as (OidcConfiguration) -> Unit
+            block(oidcConfiguration)
+            null
+        }.`when`(logtoApiClient).discover(anyOrNull())
+
+        val activity: Activity = spy(Robolectric.buildActivity(Activity::class.java).get())
+
+        val browserSignInFlow = BrowserSignInFlow(
+            logtoConfig,
+            logtoApiClient,
+            onComplete
+        )
+
+        browserSignInFlow.start(activity)
+        verify(activity).startActivity(anyOrNull())
+    }
+
+    @Test
+    fun onResultShouldCompleteWithLogtoExceptionWithInvalidUri() {
+        val browserSignInFlow = BrowserSignInFlow(
+            logtoConfig,
+            logtoApiClient,
+            onComplete,
+        )
+
+        val invalidUri: Uri = mock()
+        `when`(invalidUri.getQueryParameter(eq(QueryKey.CODE))).thenReturn(null)
+        `when`(invalidUri.getQueryParameter(eq(QueryKey.ERROR_DESCRIPTION)))
+            .thenReturn("mocked sign in error description")
+
+        browserSignInFlow.onResult(invalidUri)
+
+        verify(onComplete).invoke(logtoExceptionCaptor.capture(), tokenSetCaptor.capture())
+        assertThat(logtoExceptionCaptor.firstValue)
+            .hasMessageThat()
+            .contains("mocked sign in error description")
+        assertThat(tokenSetCaptor.firstValue).isNull()
+    }
+
+    @Test
+    fun onResultShouldCompleteWithLogtoExceptionWithEmptyUri() {
```

Comment:
wondering do we need to specify a spcific empty uri error or it should be covered by the  `CODE` or `REDIRECT_URI` validation